File: Episode12-Leaderboard/Monogame/Leaderboard.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using MonoGame.Extended;
using System;
using System.Collections.Generic;
using System.IO;

namespace Shmup
{
    internal class Leaderboard
    {
        private class ScoreData
        {
            public int Score;
            public string Name;
            public ScoreData(string line)
            {
                DecodeLine(line);
            }
            public ScoreData(string name, int score)
            {
                Name = name.Trim();
                Score = score;
            }
            private void DecodeLine(string line)
            {
                // eg line = "FRED;258"
                string[] parts = line.Split(';');
                Name = parts[0].Trim();                    // "FRED"
                Score = Convert.ToInt32(parts[1].Trim());  // 258
            }
            public string GetScoreData(string dest)
            {
                if (dest == "file")
                    return $"{Name};{Score}\n"; // "FRED;258\n"
                else
                    return $"{Name} : {Score}";
            }
        }
        public bool CurrentPlayerEntered = false;
        private List<ScoreData> scoreList = new List<ScoreData>();
        public string Name = "";
        private float PosY = 80f;
        private readonly int maxLines = 5;
        private RectangleF rect = new RectangleF(50, Shared.HEIGHT * 0.85f, Shared.WIDTH - 100f, 24);
        public Leaderboard()
        {
            PopulateScoreList();
        }
        public void AddEntry()
        {
            if (Name != "")
            {
                CurrentPlayerEntered = true;
                ScoreData scoreTemp = new ScoreData(Name, Shared.Score);
                InsertData(scoreTemp);
                Shared.InputText = "";
                WriteScoreList();
                PopulateScoreList();
            }
        }
        private void AddScoreData(string line)
        {
            // create new scoreData object from line in text file
            ScoreData scoreTemp = new ScoreData(line);
            // insert new object into correct position
            InsertData(scoreTemp);
        }
        private void InsertData(ScoreData scoredata)
        {
            int insertAt = scoreList.Count; // assume end of list
            for (int i = 0; i < scoreList.Count; i++)
            {
                if (scoredata.Score >= scoreList[i].Score) // current score > listed
                {
                    insertAt = i;
                    break;
                }
            }
            // if above loop did not find a greater score then insert at end of list
            scoreList.Insert(insertAt, scoredata);
        }
        private void PopulateScoreList()
        {
            /// Read file called HighScore.txt
            scoreList.Clear();
            if (File.Exists("Highscore.txt"))
            {
                string[] lines = File.ReadAllLines("Highscore.txt");
                foreach (string line in lines)
                {
                    if (line.Length > 0)
                        AddScoreData(line);
                }
            }
        }
        private void WriteScoreList()
        {
            using (StreamWriter outputFile = new StreamWriter("Highscore.txt"))
            {
                for(int i = 0;i < scoreList.Count; i++)
                {
                    if (i > maxLines - 1) // max 5 lines to be written
                        break;
                    outputFile.WriteLine(scoreList[i].GetScoreData("file"));
                }
            }
        }
        public void Update(KeyboardState keyboardState)
        {
            if (CurrentPlayerEntered)
            {
                if (keyboardState.IsKeyDown(Keys.C))
                {
                    Shared.GameState = Shared.GameStates["menu"];
                    CurrentPlayerEntered = false;
                    Shared.InputText = "";
                }
            }
            else
            {
                Name = Shared.InputText;
            }
        }
        public void Draw(SpriteBatch spriteBatch)
        {
            Shared.DrawString(spriteBatch, "LEADERBOARD", "size50", Shared.WIDTH * 0.5f, 0 ,Color.Yellow,"centre");    // 'Leaderboard'
            PosY = 80;
            //foreach(ScoreData score in scoreList)
            for(int i = 0; i < scoreList.Count; i++)
            {
                if (i > maxLines - 1)
                    break;
                Shared.DrawString(spriteBatch, scoreList[i].GetScoreData("display"), "size24", Shared.WIDTH * 0.5f, PosY, Color.White, "centre");
                PosY += 30;
            }
            if(CurrentPlayerEntered)    // If got input from user
            {
                Shared.DrawString(spriteBatch, "Press C to continue", "size18", 0, Shared.HEIGHT * 0.8f, Color.White, "centre");
            }
            else
            {
                Shared.DrawString(spriteBatch, "Type your name and press Enter", "size18", 0, Shared.HEIGHT * 0.8f, Color.White, "centre");
                // displayBox(SpriteBatch spriteBatch, string text, string size, RectangleF rect, Color lineColour, Color backColour, Color textColour)
                Shared.DisplayBox(spriteBatch, "NAME: " + Name, "size18", rect, Color.White, Shared.DARKBLUE, Color.White);
            }
        }
    }
}

File: Episode12-Leaderboard/Love2D/leaderboard.lua
--[[
The Monogame and Pygame versions needed a sub-class to create the 
scoreData objects. In Lua this is achieved with a simple table which
holds a list of other tables.
Each of the sub-tables holds a name/score pair:
scoreList =
{
	{name = "FRED", score = 1234},
	{name = "INKSAVER", score = 3234}
}
]]

local Class = require("lib.Class")
local Leaderboard = Class:derive("Leaderboard")
local Rectangle = require "lib.Rectangle"
local Shared = require "shared"
require "lib/Utils" -- for split function
	
function Leaderboard:addScoreData(line)
	--[[ either "FRED;356" or {"FRED", 356} ]]
	local data
	-- line could be a string or table. Lua/Python functions cannot have overloads
	-- so this method is the closest you can get
	if type(line) == "string" then
		data = line:upper():split(';')
	else -- table
		data = line
	end
	-- create a local table of name/score pairs
	local newScore = {}
	newScore.name = data[1]
	newScore.score = tonumber(data[2])
	local insertAt = #self.scoreList + 1 -- assume end of list
	-- calculate insert position high -> low
	for i = 1, #self.scoreList do --needs at least 1 item
		if newScore.score >= self.scoreList[i].score then 
			insertAt = i
			break
		end
	end
	-- insert the newScore table into scoreList
	table.insert(self.scoreList, insertAt, newScore)
end

function Leaderboard:new()
	--[[ class constructor, no parameters]]
	self.currentPlayerEntered = false
	self.scoreList = {} 		-- list of name/score sub-tables
	self.name = ""				-- get player input and store here
	self.ypos = 80
	self:populateScoreList()	-- read existing highscore text file
end

function Leaderboard:populateScoreList()
	--[[
	You have no choice where to save files. Default folder used:
	C:\Users\<username>\AppData\Roaming\LOVE\ShmupHighscore\Highscore.txt
	]]
	self.scoreList = {} -- clear the scoreList
	for line in love.filesystem.lines("Highscore.txt") do
		self:addScoreData(line)
	end
end

function Leaderboard:addEntry()
	if self.name ~= "" then
		self.currentPlayerEntered = true
		self:addScoreData({self.name, Shared.score})
		Shared.inputText = ""
		self:writeScoreList()
		self:populateScoreList()
	end
end

function Leaderboard:writeScoreList()
	--[[ C:\Users\<username>\AppData\Roaming\LOVE\ShmupHighscore\Highscore.txt
		self.scoreList = {'FRED;200','JOHN;450','MARY;286'}
	]]
	-- write the top 5 scoring values
	local text = ""
	endloop = #self.scoreList
	if endloop > 5 then endloop = 5 end
	for line = 1, endloop do
		local data = self.scoreList[line]
		text = text .. data.name .. ";".. data.score .. "\n"
	end
	success, errormsg = love.filesystem.write("Highscore.txt", text)
end

function Leaderboard:update(keyboard)
	--has Name been input?
	if self.currentPlayerEntered then -- name already obtained. waiting for c to continue
		if keyboard.isDown("c") then
			Shared.gamestate = Shared.gamestates['menu']
			self.currentPlayerEntered = false
			Shared.inputText = ""
		end
	else	-- If not get input from user
		self.name = Shared.inputText
	end
end

function Leaderboard:draw()
	Shared.drawText("LEADERBOARD", 'size50', 0, 0, "centre", Shared.YELLOW)			-- 'Leaderboard'
	-- draw a list of 5 names with current name in position related to score
	self.ypos = 80
	local lines = 0
	for _,value in ipairs(self.scoreList) do		-- up to 5 highscore names and scores
		Shared.drawText(value.name.." : "..value.score, 'size24', 0, self.ypos, "centre", Shared.WHITE)
		self.ypos = self.ypos + 30
		lines = lines + 1
		if lines > 5 then
			break
		end
	end
	
	if self.currentPlayerEntered then -- If got input from user
		Shared.drawText("Press C to continue", 'size18', 0, Shared.HEIGHT * 0.8, "centre", Shared.YELLOW)
	else
		Shared.drawText("Type your name and press Enter", 'size18', 0, Shared.HEIGHT * 0.8, "centre", Shared.WHITE)
		--displayBox(text, textSize, rect, foreColour, backColour, textColour)
		Shared.displayBox("NAME: ".. self.name, 'size18', Rectangle(50, Shared.HEIGHT * 0.85, Shared.WIDTH - 100, 24), Shared.YELLOW, Shared.DARKBLUE, Shared.WHITE)
	end
end

return Leaderboard
